Cleanup pkg/kubectl

alexandercampbell opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://github.com/kubernetes/community/blob/master/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

I was reading through `pkg/kubectl` in preparation for completing https://github.com/kubernetes/kubectl/issues/11 and noticed several opportunities for improvement. This should be easy to review since it's mostly mechanical changes. The only complicated changes are in `addFromEnvFile`, which I refactored into two functions and wrote tests for.

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
NONE
```

alexandercampbell committed:
kubectl: rename Run() -> RunRun() to clarify purpose

Run() is too overloaded in the codebase already. The other commands have
a pattern of RunExpose, RunScale, and so on. Since the command name is
"run", the associated function should be called RunRun.

alexandercampbell committed:
kubectl: rewrite docstrings in several files

Fixing inaccuracies and clarifying in the case of ambiguities.

alexandercampbell committed:
cmd: make createDeployment a private function

alexandercampbell committed:
cmd: refactor common err expr into helper function

The same redundant fmt.Sprintf() and string literal was duplicated
throughout many of the files in kubectl/cmd. Replace with a helper
function.

alexandercampbell committed:
kubectl: simplify code with help of linter

alexandercampbell committed:
kubectl: note a bug with a comment

This doesn't seem to be affecting anything and I'm not sure what the
correct behavior needs to be here. I'll highlight this in the code
review and hopefully work out a correct solution with the help of the
reviewers.

alexandercampbell committed:
kubectl: refactor addFromEnvFile, write tests

alexandercampbell commented on pkg/kubectl/stop.go:
```diff
@@ -337,6 +333,8 @@ func (reaper *StatefulSetReaper) Stop(namespace, name string, timeout time.Durat
 	if timeout == 0 {
 		numReplicas := ss.Spec.Replicas
+
+		// BUG: this timeout is never used.
```
Does anyone know if this is an actual issue? I don't know anything about this section of the code.

apelisse replied:
Yeah, I would rather ask some people if it should be removed, rather than add a comment. Did you `git-blame` and poked people?

alexandercampbell replied:
The original author of this function and unused value issue (in [30f3cb9](https://github.com/kubernetes/kubernetes/commit/30f3cb9d26fe368388502a362f9577bf20fa5d65#diff-2076b6236060a0cca0937ac2907ab263R339)) is github.com/bprashanth but he has since moved to another project.

I will make a new PR for discussion of this particular bug. For now, the comment will warn people that there might be an issue.

alexandercampbell replied:
@foxish would you mind looking at this? Your name come up as a recent editor of this function.

apelisse replied:
I wonder if line 340 below should be using `timeout` rather than `reaper.timeout` (by looking how it's done throughout the file)

foxish replied:
As discussed offline, looks like a bug here to me. The timeout should be on the individual pod teardown; it might make sense to take the user provided timeout and pass it into the scaler code.

alexandercampbell replied:
@foxish thanks for your input. I will open a separate PR that includes the fix you've described.
@apelisse let's move the discussion of this bug there. PTAL :)

mengqiy replied:
> I wonder if line 340 below should be using timeout rather than reaper.timeout (by looking how it's done throughout the file)

Maybe

I found there maybe another bug in https://github.com/alexandercampbell/kubernetes/blob/d3554ac0ce87f75e6976e97c61cc81165207aeb8/pkg/kubectl/stop.go#L294-L329. Argument `timeout` is never used in this function.

apelisse replied:
That might not be a bug though. It's implementing the `type Reaper interface`, so it needs to take the `timeout` parameter, whether it needs it or not.

alexandercampbell replied:
New PR is here: https://github.com/kubernetes/kubernetes/pull/46784

shiywang replied:
I think this command already been deperacted

alexandercampbell replied:
Yep. I've closed that PR in favor of your PR, @shiywang. I'll resolve whatever merge conflicts arise.

alexandercampbell commented:
/sig cli

alexandercampbell commented:
@kubernetes/sig-cli-bugs

fabianofranz commented:
@k8s-bot ok to test

alexandercampbell committed:
kubectl: fix inaccurate usage messages for --windows-line-endings

Part of the problem is that these are duplicated between the different
commands. I'm planning to consolidate these further.

alexandercampbell commented:
/assign pwittrock

alexandercampbell commented:
/assign @pwittrock

alexandercampbell commented:
Rebased onto master again-- ready for review

pwittrock commented:
@mengqiy FYI

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/46468)
<!-- Reviewable:end -->

apelisse commented on pkg/kubectl/doc.go:
```diff
@@ -14,7 +14,8 @@ See the License for the specific language governing permissions and
-// They are separated out into a library to support unit testing.  Most functionality should
-// be included in this package, and the main kubectl should really just be an entry point.
+// Package kubectl provides the functions used by the kubectl command line tool
+// under /cmd/. The functions are kept in this package to better support unit
```
Do you actually mean `/cmd/` or `/cmd/kubectl`? I also somehow find that first slash off: `cmd/` ?

alexandercampbell replied:
The first slash meaning from the root of the project.

apelisse replied:
Of course, I realize that. But I don't think it's a common pattern.

As an example:
```
$GOPATH/src/k8s.io/kubernetes$ cd pkg/kubectl
$GOPATH/src/k8s.io/kubernetes/pkg/kubectl$ git show HEAD:cmd/kubectl/
tree HEAD:cmd/kubectl

BUILD
OWNERS
app/
kubectl.go
$GOPATH/src/k8s.io/kubernetes/pkg/kubectl$ git show HEAD:/cmd/
fatal: Path '/cmd/' does not exist in 'HEAD'
```

alexandercampbell replied:
Ok, good call. I've corrected the comment to give an unambiguous location. Thanks!

apelisse commented on pkg/kubectl/stop.go:
```diff
@@ -48,10 +48,11 @@ const (
-// gracePeriod is time given to an API object for it to delete itself cleanly,
-// e.g., pod shutdown. It may or may not be supported by the API object.
 type Reaper interface {
+	// Stop a given given object within a namespace. timeout is how long
```
repetition: 
> Stop a given given ...

alexandercampbell replied:
Fixed, thanks!

apelisse commented on pkg/kubectl/env_file.go:
```diff
@@ -39,38 +80,23 @@ func addFromEnvFile(filePath string, addTo func(key, value string) error) error
-		}
-		// trim the line from all leading whitespace first
-		line := strings.TrimLeftFunc(string(scannedBytes), unicode.IsSpace)
 		currentLine++
```
It makes me sad that we have to keep track of how many lines we've seen, and pass that variable to `processEnvFileLine` while it's only being used once.

Should we write a `TrimBOM` function that would take the file, read the first three bytes, and `Seek` back to the beginning if it doesn't match the bom?

alexandercampbell replied:
I had actually rewritten the function for that exact purpose, but then I realized that the error message helpfully includes the line number. So I think we should still track it.

apelisse commented on pkg/kubectl/cmd/apply_edit_last_applied.go:
```diff
@@ -96,7 +96,8 @@ func NewCmdApplyEditLastApplied(f cmdutil.Factory, out, errOut io.Writer) *cobra
 	cmd.Flags().StringVarP(&options.Output, "output", "o", "yaml", "Output format. One of: yaml|json.")
-	cmd.Flags().BoolVar(&options.WindowsLineEndings, "windows-line-endings", gruntime.GOOS == "windows", "Use Windows line-endings (default Unix line-endings)")
+	cmd.Flags().BoolVar(&options.WindowsLineEndings, "windows-line-endings", runtime.GOOS == "windows",
+		"Defaults to the line ending native to your platform.")
```
While I agree that this argument is mostly obvious, it feels weird that this reads the default, but not what it actually controls.

alexandercampbell replied:
I'm not sure I follow. Can you explain a little more?
I did not change any behavior here; I only deleted the part of the help message that was blatantly incorrect :)

apelisse replied:
Sorry. You've removed "Use Windows line-endings". Now the description starts with "Default ..", which is correct, but it doesn't let you know what the parameter is actually about.

alexandercampbell replied:
The help message shows

```
      --windows-line-endings=false: Defaults to the line ending native to your platform.
```

Which I feel is unambiguous.
Do you think it would be better to re-add the words "Use Windows line endings" to the beginning of the help message?

apelisse replied:
No it's fine

droot commented on pkg/kubectl/cmd/run.go:
```diff
@@ -136,7 +136,7 @@ func addRunFlags(cmd *cobra.Command) {
 }
 
-func Run(f cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer, cmd *cobra.Command, args []string, argsLenAtDash int) error {
+func RunRun(f cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer, cmd *cobra.Command, args []string, argsLenAtDash int) error {
```
Do we really have to export this function ? I don't think these should be exported because this get wrapped in Run method of the command above.
This is applicable to all the commands in kubectl I guess ?

alexandercampbell replied:
Correct. I'd say a good majority of the exported symbols under `/kubectl/` could be made private. I did not want to create too much noise in this diff.

If I hide more of these identifiers I will isolate the rename commits to a single new PR.

droot replied:
ok, sounds good.

droot replied:
ok, sounds good.

droot commented on pkg/kubectl/env_file.go:
```diff
@@ -39,38 +80,23 @@ func addFromEnvFile(filePath string, addTo func(key, value string) error) error
-			}
+
+		if len(key) == 0 {
+			// no key means line was empty or blank
```
s/blank/a comment

alexandercampbell replied:
Good catch

droot commented on pkg/kubectl/env_file_test.go:
```diff
@@ -0,0 +1,73 @@
+			}
+		}
+	}
+}
```
+1

droot reviewed (commented):
Took a look, have a few comments.

alexandercampbell commented:
Rebased onto master again. Thanks for the great feedback. Ready for another look.

mengqiy commented on pkg/kubectl/cmd/create_service.go:
```diff
@@ -83,6 +82,10 @@ func NewCmdCreateServiceClusterIP(f cmdutil.Factory, cmdOut io.Writer) *cobra.Co
 	return cmd
 }
 
+func errUnsupportedGenerator(cmd *cobra.Command, generatorName string) error {
```
Maybe we should move this helper function out to util.

alexandercampbell replied:
It's only used by the functions in this package, so I kept it as an unexported function. I'm trying to avoid cross-package dependencies where it's not absolutely necessary :)
If you feel strongly on the issue I will move it, no problem.

shiywang replied:
I think it's not necessary write a new function which only use in this package, or like @mengqiy suggest move this helper function out to util

alexandercampbell replied:
@mengqiy what do you think? Move to `kubectl/util`?
I have not done this yet because the function is used only within the package.

alexandercampbell replied:
> I think it's not necessary to write a new function which only use in this package

It's better than copying and pasting the error message everywhere :)

mengqiy commented on pkg/kubectl/env_file_test.go:
```diff
@@ -0,0 +1,73 @@
+		{"key is returned even with no value",
+			[]byte{' ', 'x', '='}, 0, "x", "", ""},
+
+		{"a key is returned even with no value next to the equals sign",
```
What is the difference between this one and the one above it.

mengqiy commented on pkg/kubectl/env_file_test.go:
```diff
@@ -0,0 +1,73 @@
+		{"a key is returned even with no value next to the equals sign",
+			[]byte{' ', 'x', '='}, 0, "x", "", ""},
+
+		{"a key is returned even with no value and no equals sign",
```
The title seems not accurate.

alexandercampbell replied:
Good call! I have deleted the two tests with issues and written a new test function that tests the "no equals sign" case. This particular case reads from the environment and I felt it was important to isolate that behavior.

mengqiy commented:
Don't forget to squash the commits to a reasonable number, 17 is too many.

apelisse commented:
> Don't forget to squash the commits to a reasonable number before merging, 17 is too many.

I disagree with that. Squashing "fix typo" commits is useful, but these are legitimate, independent, well organized commits.

alexandercampbell commented:
@mengqiy I have rebased the commits and squashed a few closely-related commits. There are now 9 commits. Does that strike a good balance? I'm trying to apply the Single Responsibility Principle to commits.

Good call with the test cases. I have updated them to have accurate names and better coverage.
Ready for another look.

falenn commented:
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  curtis.l.bates@gmail.com
    Domain imwiz.com has exceeded the max emails per hour (154/150 (102%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from o5.sgmail.github.com ([192.254.113.10]:25095)
	by box969.bluehost.com with esmtps (TLSv1.2:ECDHE-RSA-AES128-GCM-SHA256:128)
	(Exim 4.87)
	(envelope-from <bounces+848413-5c7e-dev=imwiz.com@sgmail.github.com>)
	id 1dGTKo-000ThR-6h
	for dev@imwiz.com; Thu, 01 Jun 2017 10:52:32 -0600
DKIM-Signature: v=1; a=rsa-sha1; c=relaxed/relaxed; d=github.com; 
	h=from:reply-to:to:cc:in-reply-to:references:subject:mime-version:content-type:content-transfer-encoding:list-id:list-archive:list-post:list-unsubscribe; 
	s=s20150108; bh=qyuGA190Lo9oiv1w7Xe9/gmc3qQ=; b=G6sdmG0ExmhjRtxI
	p7kIEVyxQR/EP1t0crpC9jzeV7cgL+tLQ8ebCmIYv0PmrPDGOkJdiI78QneTfch8
	dwMVPuBW8VRHBoT+Sh5KIGE4RQ+zkCL59iQnUknyx+Wlzmz8vDhRndn3Vo7/Ewbs
	yfrjR4p6dxVTIx+ZkqZkSIS862U=
Received: by filter0469p1mdw1.sendgrid.net with SMTP id filter0469p1mdw1-30694-5930460F-68
        2017-06-01 16:51:27.93130428 +0000 UTC
Received: from github-smtp2a-ext-cp1-prd.iad.github.net (github-smtp2a-ext-cp1-prd.iad.github.net [192.30.253.16])
	by ismtpd0001p1iad1.sendgrid.net (SG) with ESMTP id 0lXuyth9SI2Phla4ioqxog
	for <dev@imwiz.com>; Thu, 01 Jun 2017 16:51:27.866 +0000 (UTC)
Date: Thu, 01 Jun 2017 09:51:27 -0700
From: Alexander Campbell <notifications@github.com>
Reply-To: kubernetes/kubernetes <reply@reply.github.com>
To: kubernetes/kubernetes <kubernetes@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Message-ID: <kubernetes/kubernetes/pull/46468/c305553642@github.com>
In-Reply-To: <kubernetes/kubernetes/pull/46468@github.com>
References: <kubernetes/kubernetes/pull/46468@github.com>
Subject: Re: [kubernetes/kubernetes] Cleanup pkg/kubectl (#46468)
Mime-Version: 1.0
Content-Type: multipart/alternative;
 boundary="--==_mimepart_5930460fbcde8_3b653ffce3591c3878592";
 charset=UTF-8
Content-Transfer-Encoding: 7bit
Precedence: list
X-GitHub-Sender: alexandercampbell
X-GitHub-Recipient: falenn
X-GitHub-Reason: subscribed
List-ID: kubernetes/kubernetes <kubernetes.kubernetes.github.com>
List-Archive: https://github.com/kubernetes/kubernetes
List-Post: <mailto:reply@reply.github.com>
List-Unsubscribe: <mailto:unsub+000ab60a0783ac8e26b4f13fd518e5e2b49125b2acfd2dfc92cf000000011548080f92a169ce0dcbddd1@reply.github.com>,
 <https://github.com/notifications/unsubscribe/AAq2CtfhdcikdVA8aGhN_bpqNbqEpGn8ks5r_uwPgaJpZM4Nm6WO>
X-Auto-Response-Suppress: All
X-GitHub-Recipient-Address: dev@imwiz.com
X-SG-EID: APO41b8ovafPb3SK9rw3vL8cZ+ONyTJ/Vh2Npzd11QIXhXymRxp9nJ4jTK4SMdO1J4ClpJapLiJCgE
 uYDaQPxnkBxhIllUUDDO6Zzn1NLJtrvsztgQqbX9uLksBzqVgJZ02/b4N+wXyfEG9K6xFy9qkLc6LA
 8VBgj60nDAvxfINP/WdNSBaINgQESUb/Ko+PrLz9LsfSDgQstBiMKDXZaRRz/CevblTkgUnY2oLXn+
 Q=
X-Spam-Status: No, score=-2.7
X-Spam-Score: -26
X-Spam-Bar: --
X-Ham-Report: Spam detection software, running on the system "box969.bluehost.com",
 has NOT identified this incoming email as spam.  The original
 message has been attached to this so you can view it or label
 similar future email.  If you have any questions, see
 root\@localhost for details.
 
 Content preview:  @mengqiy I have rebased the commits and squashed a few closely-related
    commits. There are now 9 commits. Does that strike a good balance? I'm trying
    to apply the Single Responsibility Principle to commits. [...] 
 
 Content analysis details:   (-2.7 points, 4.0 required)
 
  pts rule name              description
 ---- ---------------------- --------------------------------------------------
 -0.0 RP_MATCHES_RCVD        Envelope sender domain matches handover relay domain
 -2.8 RCVD_IN_MSPIKE_H2      RBL: Average reputation (+2)
                             [192.254.113.10 listed in wl.mailspike.net]
 -0.0 SPF_PASS               SPF: sender matches SPF record
  0.7 HTML_IMAGE_ONLY_28     BODY: HTML: images with 2400-2800 bytes of words
  0.0 HTML_MESSAGE           BODY: HTML included in message
 -0.1 DKIM_VALID_AU          Message has a valid DKIM or DK signature from author's
                             domain
 -0.1 DKIM_VALID             Message has at least one valid DKIM or DK signature
  0.1 DKIM_SIGNED            Message has a DKIM or DK signature, not necessarily valid
 -0.6 AWL                    AWL: Adjusted score from AWL reputation of From: address
X-Spam-Flag: NO

----==_mimepart_5930460fbcde8_3b653ffce3591c3878592
Content-Type: text/plain;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

@mengqiy I have rebased the commits and squashed a few closely-related commits. There are now 9 commits. Does that strike a good balance? I'm trying to apply the Single Responsibility Principle to commits.

Good call with the test cases. I have updated them to have accurate names and better coverage.
Ready for another look.

-- 
You are receiving this because you are subscribed to this thread.
Reply to this email directly or view it on GitHub:
https://github.com/kubernetes/kubernetes/pull/46468#issuecomment-305553642
----==_mimepart_5930460fbcde8_3b653ffce3591c3878592
Content-Type: text/html;
 charset=UTF-8
Content-Transfer-Encoding: quoted-printable

<p><a href=3D"https://github.com/mengqiy" class=3D"user-mention">@mengqiy</=
a> I have rebased the commits and squashed a few closely-related commits. T=
here are now 9 commits. Does that strike a good balance? I'm trying to appl=
y the Single Responsibility Principle to commits.</p>
<p>Good call with the test cases. I have updated them to have accurate name=
s and better coverage.<br>
Ready for another look.</p>

<p style=3D"font-size:small;-webkit-text-size-adjust:none;color:#666;">&mda=
sh;<br />You are receiving this because you are subscribed to this thread.<=
br />Reply to this email directly, <a href=3D"https://github.com/kubernetes=
/kubernetes/pull/46468#issuecomment-305553642">view it on GitHub</a>, or <a=
 href=3D"https://github.com/notifications/unsubscribe-auth/AAq2CovwryhZq-lw=
rRseyaR8AG2EKltIks5r_uwPgaJpZM4Nm6WO">mute the thread</a>.<img alt=3D"" hei=
ght=3D"1" src=3D"https://github.com/notifications/beacon/AAq2CnPNuhBxhgUtrz=
irARZ8dTzsE0cFks5r_uwPgaJpZM4Nm6WO.gif" width=3D"1" /></p>
<div itemscope itemtype=3D"http://schema.org/EmailMessage">
<div itemprop=3D"action" itemscope itemtype=3D"http://schema.org/ViewAction=
">
  <link itemprop=3D"url" href=3D"https://github.com/kubernetes/kubernetes/p=
ull/46468#issuecomment-305553642"></link>
  <meta itemprop=3D"name" content=3D"View Pull Request"></meta>
</div>
<meta itemprop=3D"description" content=3D"View this Pull Request on GitHub"=
></meta>
</div>

<script type=3D"application/json" data-scope=3D"inboxmarkup">{"api_version"=
:"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"Gi=
tHub"},"entity":{"external_key":"github/kubernetes/kubernetes","title":"kub=
ernetes/kubernetes","subtitle":"GitHub repository","main_image_url":"https:=
//cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95f=
c-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com=
/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":=
{"name":"Open in GitHub","url":"https://github.com/kubernetes/kubernetes"}}=
,"updates":{"snippets":[{"icon":"PERSON","message":"@alexandercampbell in #=
46468: @mengqiy I have rebased the commits and squashed a few closely-relat=
ed commits. There are now 9 commits. Does that strike a good balance? I'm t=
rying to apply the Single Responsibility Principle to commits.\r\n\r\nGood =
call with the test cases. I have updated them to have accurate names and be=
tter coverage.\r\nReady for another look."}],"action":{"name":"View Pull Re=
quest","url":"https://github.com/kubernetes/kubernetes/pull/46468#issuecomm=
ent-305553642"}}}</script>=

----==_mimepart_5930460fbcde8_3b653ffce3591c3878592--

mengqiy commented:
/lgtm

mengqiy commented:
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

alexandercampbell commented:
Rebased. PTAL :)

krzyzacy commented:
/retest

shiywang commented on pkg/kubectl/cmd/create.go:
```diff
@@ -40,7 +40,7 @@ type CreateOptions struct {
 
 var (
 	create_long = templates.LongDesc(i18n.T(`
```
nit: how about also change coding style here ? also I'm not sure if there's any other places exist coding style like this in kubectl

alexandercampbell replied:
Good call! Updated.

alexandercampbell commented:
[Failing unit test](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46468/pull-kubernetes-unit/37518/) appears to be a flake since it's passing for me locally. Rebasing onto latest master to see if it continues to fail.

alexandercampbell commented:
@mengqiy would you mind looking at this again? Your LGTM was removed when I fixed merge conflicts.

pwittrock commented:
/lgtm

pwittrock commented:
/approve

alexandercampbell commented:
Thanks @pwittrock!
